### PR TITLE
Add LLDPProcessed information from ironic-inspector data

### PR DESCRIPTION
**Pull request by shibaPuppy**
Collect node LLDP information.

ex) if there is no LLDP information
```
apiVersion: metal3.io/v1alpha1
kind: BareMetalHost
metadata:
  name: node-3
  namespace: metal3
spec:
  automatedCleaningMode: metadata
  bmc:
    address: ipmi://192.168.111.1:6233
    credentialsName: node-3-bmc-secret
  bootMACAddress: 00:cc:0f:73:1b:fb
  bootMode: legacy
  online: true
status:
.....
    nics:
    - ip: 192.168.111.49
      lldp: {}
      mac: 00:cc:0f:73:1b:fd
      model: 0x1af4 0x0001
      name: enp2s0
    - ip: fe80::2cc:fff:fe73:1bfd%enp2s0
      lldp: {}
      mac: 00:cc:0f:73:1b:fd
      model: 0x1af4 0x0001
      name: enp2s0
    - ip: 172.22.0.83
      lldp: {}
      mac: 00:cc:0f:73:1b:fb
      model: 0x1af4 0x0001
      name: enp1s0
      pxe: true
    - ip: fe80::2cc:fff:fe73:1bfb%enp1s0
      lldp: {}
      mac: 00:cc:0f:73:1b:fb
      model: 0x1af4 0x0001
      name: enp1s0
      pxe: true
    ramMebibytes: 8192
....
```

### CODE REVIEW

File: apis/metal3.io/v1alpha1/baremetalhost_types.go [hunk 1]
```diff
@@ -599,6 +630,9 @@ type NIC struct {
 
 	// Whether the NIC is PXE Bootable
 	PXE bool `json:"pxe,omitempty"`
+
+	// Save ironic inspection data to lldp
+	LLDPProcessed LLDPProcessed `json:"lldp,omitempty"`
```
Comment (dtantsur) [inline apis/metal3.io/v1alpha1/baremetalhost_types.go:635]: The "Processes" part is essentially an internal detail, let's leave just LLDP.

File: apis/metal3.io/v1alpha1/baremetalhost_types.go [hunk 2]
```diff
@@ -564,6 +564,37 @@ type Storage struct {
 // +kubebuilder:validation:Maximum=4094
 type VLANID int32
 
+type LLDPProcessed struct {
+	// https://github.com/openstack/python-ironic-inspector-client/blob/master/ironic_inspector_client/resource.py#L24-L59
```
Comment (dtantsur) [inline apis/metal3.io/v1alpha1/baremetalhost_types.go:568]: This becomes a comment from the Interface field, which is wrong. Ideally, we need a meaningful comment for each item, but definitely not a link to the code.

File: apis/metal3.io/v1alpha1/baremetalhost_types.go [hunk 3]
```diff
@@ -564,6 +564,37 @@ type Storage struct {
...
+	Interface                         string           `json:"interface,omitempty"`
+	Mac                               string           `json:"mac,omitempty"`
+	NodeIdent                         string           `json:"node_ident,omitempty"`
+	SwitchCapabilitiesEnabled         []string         `json:"switch_capabilities_enabled,omitempty"`
```
Comment (dtantsur) [inline apis/metal3.io/v1alpha1/baremetalhost_types.go:572]: I wonder if we should create a Switch substructure just to make the API a bit more usable.

File: apis/metal3.io/v1alpha1/baremetalhost_types.go [hunk 4]
```diff
@@ -564,6 +564,37 @@ type Storage struct {
...
+	SwitchCapabilitiesSupport         []string         `json:"switch_capabilities_support,omitempty"`
+	SwitchChassisId                   string           `json:"switch_chassis_id,omitempty"`
+	SwitchMgmtAddresses               []string         `json:"switch_mgmt_addresses,omitempty"`
+	SwitchPortAutonegotiation_enabled string           `json:"switch_port_autonegotiation_enabled,omitempty"`
```
Comment (dtantsur) [inline apis/metal3.io/v1alpha1/baremetalhost_types.go:576]: Let's not mix cases.

File: apis/metal3.io/v1alpha1/baremetalhost_types.go [hunk 5]
```diff
@@ -564,6 +564,37 @@ type Storage struct {
...
+	SwitchPortId                      string           `json:"switch_port_id,omitempty"`
+	SwitchPortLinkAggregationEnabled  bool             `json:"switch_port_link_aggregation_enabled,omitempty"`
+	SwitchPortLinkAggregationId       int              `json:"switch_port_link_aggregation_id,omitempty"`
+	SwitchPortLinkAggregationSupport  bool             `json:"switch_port_link_aggregation_support,omitempty"`
+	SwitchPortManagementVlanId        string           `json:"switch_port_management_vlan_id,omitempty"`
+	SwitchPortMauType                 string           `json:"switch_port_mau_type,omitempty"`
+	SwitchPortMTU                     int              `json:"switch_port_mtu,omitempty"`
+	SwitchPortPhysicalCapabilities    string           `json:"switch_port_physical_capabilities,omitempty"`
+	SwitchPortProtocolVlanEnabled     string           `json:"switch_port_protocol_vlan_enabled,omitempty"`
+	SwitchPortProtocolVlanIds         string           `json:"switch_port_protocol_vlan_ids,omitempty"`
```
Comment (dtantsur) [inline apis/metal3.io/v1alpha1/baremetalhost_types.go:588]: It's weird to have VlanIds as `string` in one place and as `[]string` (why not int?) in another. I don't remember why it was done initially, but I suggest we clean it up.

File: pkg/provisioner/ironic/hardwaredetails/hardwaredetails.go [hunk 6]
```diff
@@ -24,6 +28,21 @@ func GetHardwareDetails(data *introspection.Data) *metal3v1alpha1.HardwareDetail
 	return details
 }
 
+func getLLDPProcessed(intf introspection.BaseInterfaceType) metal3v1alpha1.LLDPProcessed {
+	lldpjson, err := json.Marshal(intf.LLDPProcessed)
```
Comment (dtantsur) [inline pkg/provisioner/ironic/hardwaredetails/hardwaredetails.go:32]: I hope there is a more efficient way to get the structure than to serialize it and deserialize back...

**Review by dtantsur — CHANGES_REQUESTED** [review]
/ok-to-test

I've left some comments, but I think it's a significant enough API addition to justify a spec in metal3-docs first.

File: pkg/provisioner/ironic/hardwaredetails/hardwaredetails.go (hunk 6 above)
Comment (shibaPuppy) [reply pkg/provisioner/ironic/hardwaredetails/hardwaredetails.go:32]: @dtantsur 
i can't think of a good way.
do you have any code to suggest?

Comment (shibaPuppy) [reply pkg/provisioner/ironic/hardwaredetails/hardwaredetails.go:32]: instead of dump the entiry data, how about saving only specific values ​​first?

**Comment by furkatgofurov7** [comment]
Please label the PR appropriately based on https://github.com/metal3-io/baremetal-operator/blob/main/CONTRIBUTING.md#contributing-a-patch to make it easier to sort it out while releasing in the changelog, thanks!

**Comment by dtantsur** [comment]
On a second thought... maybe add a new CR for LLDP information and link to it? My worry is that this change will make hardware data absolutely unreadable in case when LLDP data is actually present.

**Comment by shibaPuppy** [comment]
@Rozzii @dtantsur @furkatgofurov7 
Doesn't anyone work on the lldp controller?

If no one does it, I'll try to start next month.

**Comment by Rozzii** [comment]
> @Rozzii @dtantsur @furkatgofurov7 Doesn't anyone work on the lldp controller?
> 
> If no one does it, I'll try to start next month.

Hello,
I believe there is no one working on that as far as I know.
It would be a great help if you would work on that :D !!!

**Comment by Rozzii** [comment]
@shibaPuppy Should we let this PR get closed? I assume you are working on the new CR.

**Comment by shibaPuppy** [comment]
@Rozzii @dtantsur @furkatgofurov7 
https://github.com/metal3-io/baremetal-operator/blob/main/apis/metal3.io/v1alpha1/baremetalhost_types.go#L669-L677

would it be better to create a CR in the form of "NetworkDiscoveryData" referenced in "HardwareDetail" and add it?
I'm curious about your opinion.


ex)
```suggeust
type HardwareDetails struct {
	SystemVendor HardwareSystemVendor `json:"systemVendor,omitempty"`
	Firmware     Firmware             `json:"firmware,omitempty"`
	RAMMebibytes int                  `json:"ramMebibytes,omitempty"`
	NIC          []NIC                `json:"nics,omitempty"`
	Storage      []Storage            `json:"storage,omitempty"`
	CPU          CPU                  `json:"cpu,omitempty"`
	Hostname     string               `json:"hostname,omitempty"`
	NetworkDiscoveryData *NetworkDiscoveryData `json: "networkdiscoverydata,omitempty"`
}
```

**Comment by Rozzii** [comment]
I see there is still interest in this so I will reopen.

**Comment by Rozzii** [comment]
> @Rozzii @dtantsur @furkatgofurov7 https://github.com/metal3-io/baremetal-operator/blob/main/apis/metal3.io/v1alpha1/baremetalhost_types.go#L669-L677
> 
> would it be better to create a CR in the form of "NetworkDiscoveryData" referenced in "HardwareDetail" and add it? I'm curious about your opinion.
> 
> ex)
> 
> ```
> type HardwareDetails struct {
> 	SystemVendor HardwareSystemVendor `json:"systemVendor,omitempty"`
> 	Firmware     Firmware             `json:"firmware,omitempty"`
> 	RAMMebibytes int                  `json:"ramMebibytes,omitempty"`
> 	NIC          []NIC                `json:"nics,omitempty"`
> 	Storage      []Storage            `json:"storage,omitempty"`
> 	CPU          CPU                  `json:"cpu,omitempty"`
> 	Hostname     string               `json:"hostname,omitempty"`
> 	NetworkDiscoveryData *NetworkDiscoveryData `json: "networkdiscoverydata,omitempty"`
> }
> ```

This PR has passed is a bit out of date now let's organize this discussion a bit better.
Please create an Issue in BMO related to LLDP then let's continue the discussion there, also feel free to create a proposal about the "NetworkDiscoveryData" in the docs repo @shibaPuppy

**Pull request closed without merging**